src/state.rs

```diff
@@ -18,140 +22,208 @@ pub struct OPRFServer {
-/// advance and key rotation according to the given Config.
-#[instrument(skip_all)]
-pub async fn epoch_loop(state: OPRFState, config: &Config) {
-    let epochs = config.first_epoch..=config.last_epoch;
-
-    let interval =
-        std::time::Duration::from_secs(config.epoch_seconds.into());
-    info!("rotating epoch every {} seconds", interval.as_secs());
-
-    let start_time = time::OffsetDateTime::now_utc();
-    // Epoch base_time comes from a config argument if given,
-    // otherwise use start_time.
-    let base_time = config.epoch_base_time.unwrap_or(start_time);
-    info!(
-        "epoch base time = {}",
-        base_time
-            .format(&Rfc3339)
-            .expect("well-known timestamp format should always succeed")
-    );
-
-    // Calculate where we are in the epoch schedule relative to the
-    // base time. We may need to start in the middle of the range.
-    assert!(
-        start_time >= base_time,
-        "epoch-base-time should be in the past"
-    );
-    // The time difference will be positive after the assert.
-    // The ratio of two Durations is an f64 (in seconds) which covers
-    // the representable range of `OffsetDateTime`.
-    let elapsed_epochs = ((start_time - base_time) / interval).floor() as u64;
-
-    // The `epochs` range is `u8`, so the length can be no more
-    // than `u8::MAX + 1`, making it safe to truncate the modulo.
-    let offset = elapsed_epochs % epochs.len() as u64;
-    let current_epoch = epochs.start() + offset as u8;
-
-    // Advance to the current epoch if base time indicates we started
-    // in the middle of a sequence.
-    if current_epoch != config.first_epoch {
-        info!(
-            "Puncturing obsolete epochs {}..{} to match base time",
-            config.first_epoch, current_epoch
-        );
-        let mut s = state.write().expect("Failed to lock OPRFState");
-        for epoch in config.first_epoch..current_epoch {
-            s.server
-                .puncture(epoch)
-                .expect("Failed to puncture obsolete epoch");
+/// Container for OPRF instances
+pub struct OPRFServer {
+    /// All OPRF instances, keyed by instance name
+    pub instances: HashMap<String, RwLock<OPRFInstance>>,
+    /// The name of the default instance
+    pub default_instance: String,
+}
+
+/// Arc wrapper for OPRFServer
+pub type OPRFState = Arc<OPRFServer>;
+
+struct StartingEpochInfo {
+    elapsed_epoch_count: usize,
+    next_rotation: OffsetDateTime,
+}
+
+impl StartingEpochInfo {
+    fn calculate(base_time: OffsetDateTime, instance_epoch_duration: CalendarDuration) -> Self {
+        let now = time::OffsetDateTime::now_utc();
+        let mut elapsed_epoch_count = 0;
+        let mut next_rotation = base_time + instance_epoch_duration;
+        while next_rotation < now {
+            next_rotation = next_rotation + instance_epoch_duration;
+            elapsed_epoch_count += 1;
+        }
```

Comment:
Hmm. If our goal is to align with periodic points on the calendar, maybe it would be less convoluted to have a `Daily, Weekly, Monthly` enum and calculate the delay to the next point from that, instead of relying on the right interaction with the base time?

The `time::Date` type has `next_day()` and `next_occurance(Weekday)` methods that might be helpful here. Looks like it doesn't provide exactly what we need for months, but the `Month` type has `next()` and `previous()` methods which wrap around properly, so a value could be constructed that way.

Reply:
>Hmm. If our goal is to align with periodic points on the calendar, maybe it would be less convoluted to have a Daily, Weekly, Monthly enum and calculate the delay to the next point from that, instead of relying on the right interaction with the base time?

i'm unsure how using an enum would simplify the calculation of the initial epoch, which was the code that you highlighted. previously we used arithmetic to calculate the starting epoch, but it seems that we can no longer do this with monthly epochs due to the fact that months vary in the amount of days. i switched to an iterative approach to account for this.

Also, switching from a duration string defined by the user to an enum will make testing short epoch duration difficult, which was a concern that you mentioned here: https://github.com/brave/star-randsrv/pull/50#discussion_r1154779384

The ability to define a custom epoch duration provides this kind of flexibility.

> The time::Date type has next_day() and next_occurance(Weekday) methods that might be helpful here. Looks like it doesn't provide exactly what we need for months, but the Month type has next() and previous() methods which wrap around properly, so a value could be constructed that way.

The new small crate `calendar-duration` provides an abstraction layer for this.

Reply:
> i'm unsure how using an enum would simplify the calculation of the initial epoch

Sorry, I don't mean that it simplifies the code. I meant an enum would simplify the interface. This change does two things. It adds a parser so someone using the command line doesn't have to figure out how many seconds a week is, or whatever. That's nice. Separately, it allows specifying durations over a month in a way that aligns with calendar dates, which isn't something that was possible before, because months aren't a fixed number of seconds.

But, to get what we want, which is to have epochs rotating at the zero hour UTC on daily, weekly, and monthly cadences, we still have to think carefully about whether the `--epoch-base-time` we pass is a Monday and the first day of a month. ([It is](https://github.com/brave/star-randsrv/blob/bb7aa6b466cae93d7755a5535aba3e1d1424a3f2/start.sh#L19). Did you do that on purpose?) Seems like it would be nicer to be able to just request that directly and let the code do the math.

> an enum will make testing short epoch duration difficult

I think we could still include a `Seconds(u64)` variant in the enum for testing or other custom intervals.

Also, would this change support rotating epochs more frequently than the calendar measurement period like we've discussed doing with P3A? What arguments would I pass to rotate three times a month?

Reply:
> ([It is](https://github.com/brave/star-randsrv/blob/bb7aa6b466cae93d7755a5535aba3e1d1424a3f2/start.sh#L19). Did you do that on purpose?)

nope, that was all you :smiley: 

> I think we could still include a Seconds(u64) variant in the enum for testing or other custom intervals.

what benefit would this provide over just specifying `--epoch-duration 30s` or something similar?

> Also, would this change support rotating epochs more frequently than the calendar measurement period like we've discussed doing with P3A? What arguments would I pass to rotate three times a month?

This is a good question. I'm afraid not. One idea for a future PR: we could rename `--epoch-duration` to `--report-cadence-duration` and add a switch called `--epochs-per-report-period`. we could perform a small amount of extra arithmetic to calculate the epoch; I expect that the burden wouldn't be significant